crates/wasm-mutate/src/mutators/peephole/eggsy/encoder/mod.rs

```diff
@@ -0,0 +1,549 @@
+//! Helper methods for encoding eterm expressions to Wasm and back
+
+use crate::mutators::peephole::dfg::MiniDFG;
+use crate::mutators::peephole::eggsy::encoder::expr2wasm::expr2wasm;
+use crate::mutators::peephole::{Lang, EG};
+use crate::{
+    mutators::peephole::{
+        dfg::{BBlock, StackEntry},
+        OperatorAndByteOffset,
+    },
+    ModuleInfo,
+};
+use egg::RecExpr;
+
+use wasm_encoder::{Function, Instruction, MemArg};
+
+pub mod expr2wasm;
+pub mod rebuild;
+
+/// Turns wasm to eterm and back
+pub struct Encoder;
+
+/// Traversing node events
+enum TraversalEvent {
+    Enter,
+    Exit,
+}
+
+impl Encoder {
+    fn writestackentry(
+        dfg: &MiniDFG,
+        entry: &StackEntry,
+        newfunc: &mut Function,
+    ) -> crate::Result<()> {
+        let mut worklist = vec![
+            (entry, TraversalEvent::Exit),
+            (entry, TraversalEvent::Enter),
+        ];
+
+        while let Some((entry, event)) = worklist.pop() {
+            match event {
+                TraversalEvent::Enter => {
+                    // push operands
+                    for idx in entry.operands.iter().rev() {
+                        let entry = &dfg.entries[*idx];
+                        worklist.push((entry, TraversalEvent::Exit));
+                        worklist.push((entry, TraversalEvent::Enter));
+                    }
+                }
+                TraversalEvent::Exit => {
+                    match &entry.operator {
+                        Lang::I32Add(_) => {
+                            newfunc.instruction(&Instruction::I32Add);
+                        }
+                        Lang::I64Add(_) => {
+                            newfunc.instruction(&Instruction::I64Add);
+                        }
+                        Lang::I32Sub(_) => {
+                            newfunc.instruction(&Instruction::I32Sub);
```

Comment:
Wait -- how is this different from `expr2wasm`? Why do we have two versions of `Lang` to Wasm encoding?

Reply:
Working on it, after removing our "intermediate" `StackEntry` conversion a lot of code start to be repetitive, and this case is probably something that we can remove.